book/src/qobject/cpp_object.md

```diff
@@ -5,36 +5,106 @@ SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+  pub struct MyObject {}
+}
 ```
 
-If there is a [`Signals` enum](./signals_enum.md) then you can call `emit_queued(&mut self, Signals)` on the `CppObj` to emit a signal.
+``` rust,ignore,noplayground
+// In another module
+fn my_function(parameter: &crate::qt_types::qobject::MyObject) { /* ... */ }
+```
+
+Unfortunately, nested QObjects aren't yet supported by CXX-Qt.
+Progress can be tracked in [#299](https://github.com/KDAB/cxx-qt/issues/299).
+
+## `impl qobject::T`
+As mentioned before, the C++ QObject is exposed to Rust as an opaque CXX type.
+A side-effect of this is that you can actually extend this type with your own functionality by using normal `impl` blocks.
+Because `qobject::T` is an opaque C++ type, the same rules as with normal CXX opaque types apply.
+Most importantly this means that the type may never be accessed by value or by mutable self reference directly.
+Rather, all methods must use either `&self` or `self: Pin<&mut Self>` as the self types.
 
-Note that signals are queued to avoid a deadlock if the `Q_EMIT` is `Qt::DirectConnection` connected to a Rust invokable on the same QObject that has caused the `Q_EMIT`, as this would then try to lock the internal Rust object which is already locked.
+For more information about pinning, see: https://doc.rust-lang.org/std/pin/
 
-```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/rust/src/signals.rs:book_rust_obj_impl}}
+In addition to methods that extend the Rust interface of a `qobject::T`, you may also mark methods within an `impl qobject::T` block with the `#[qinvokable]` attribute.
+These methods will then be exposed to the C++ QObject and can then be called by C++ or QML.
+See the [QObject page](./qobject_struct.md#invokables) for more details.
+
+## Available Functions
+
+Every `qobject::T` struct provides the following methods:
+
+### Access to the Qt thread
+``` rust
+fn qt_thread(&self) -> UniquePtr<CxxQtThread>
 ```
+This function provides you with a handle to the Qt thread the QObject resides in.
+This is helpful as the QObject itself is neither Send, nor Sync.
+The CxxQtThread however is Send and can therefore be moved into a different thread.
+By using the `queue` function on the CxxQtThread, you may then queue a Rust lambda onto the Qt thread again.
+The lambda also takes a pinned mutable reference to the QObject, so that it can modify it.
 
-## Threading
+See the [CxxQtThread page](./cxxqtthread.md) for more details.
 
-`CppObj` can be used for [threading](../concepts/threading.md) to access the [`CxxQtThread<T>`](./cxxqtthread.md) via the `qt_thread(&self)` method.
+### Signal emission
+``` rust
+fn emit(self: Pin<&mut Self>, signal: /*Your Signals enum goes here*/)
+```
+If there is a [Signals enum](./signals_enum.md) defined, CXX-Qt will generate the appropriate `emit` function to allow you to emit signals.
 
-## Type Wrappers
+See the [Signals enum page](./signals_enum.md) for more details.
+
+### Access to internal Rust struct
```

Comment:
Should this section mention the helpers like `set_property` and the `set_field` (for non properties?) (and prop/prop_mut etc). As a field that is not a property has a safe `field_mut` whereas the property one is unsafe etc. And should this link to the qobject_struct page.

Reply:
I've added links to the properties, as well as "private fields & methods" sections.